src/folding/nova/ivc.rs

```diff
@@ -0,0 +1,272 @@
+use ark_crypto_primitives::sponge::{poseidon::PoseidonConfig, Absorb};
+use ark_ec::{CurveGroup, Group};
+use ark_ff::PrimeField;
+use ark_relations::r1cs::ConstraintSynthesizer;
+use ark_relations::r1cs::ConstraintSystem;
+use ark_std::rand::Rng;
+use ark_std::{One, Zero};
+use core::marker::PhantomData;
+
+use super::circuits::{AugmentedFCircuit, FCircuit};
+use super::{
+    nifs::NIFS,
+    traits::{NovaR1CS, NovaTranscript},
+    CommittedInstance, Witness,
+};
+use crate::ccs::r1cs::R1CS;
+use crate::frontend::arkworks::{extract_r1cs, extract_z}; // TODO once Frontend trait is ready, use that
+use crate::pedersen::{Params as PedersenParams, Pedersen};
+use crate::transcript::Transcript;
+use crate::{unwrap_or_return_err, Error};
+
+pub struct IVC<C1, C2, FC, Tr>
+where
+    C1: CurveGroup,
+    C2: CurveGroup,
+    FC: FCircuit<C1::ScalarField>,
+    Tr: Transcript<C1> + NovaTranscript<C1>,
+{
+    _c2: PhantomData<C2>,
+    r1cs: R1CS<C1::ScalarField>,
+    pub poseidon_config: PoseidonConfig<C1::ScalarField>,
+    pub pedersen_params: PedersenParams<C1>,
+    pub F: FC, // F circuit
+    pub transcript: Tr,
+    i: C1::ScalarField,
+    z_0: Vec<C1::ScalarField>,
+    z_i: Vec<C1::ScalarField>,
+    w_i: Witness<C1>,
+    u_i: CommittedInstance<C1>,
+    W_i: Witness<C1>,
+    U_i: CommittedInstance<C1>,
+}
+
+impl<C1, C2, FC, Tr> IVC<C1, C2, FC, Tr>
+where
+    C1: CurveGroup,
+    C2: CurveGroup,
+    FC: FCircuit<C1::ScalarField>,
+    Tr: Transcript<C1> + NovaTranscript<C1>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+{
+    pub fn new<R: Rng>(
+        rng: &mut R,
+        transcript_config: Tr::TranscriptConfig,
+        poseidon_config: PoseidonConfig<C1::ScalarField>,
+        F: FC,
+        z_0: Vec<C1::ScalarField>,
+    ) -> Result<Self, Error> {
+        // prepare the circuit to obtain its R1CS
+        let cs = ConstraintSystem::<C1::ScalarField>::new_ref();
+
+        let augmented_F_circuit = AugmentedFCircuit::<C1, FC>::empty(&poseidon_config, F);
+
+        augmented_F_circuit.generate_constraints(cs.clone())?;
+        cs.finalize();
+        let cs = unwrap_or_return_err!(cs.into_inner(), Err(Error::NoInnerConstraintSystem));
+        let r1cs = extract_r1cs::<C1::ScalarField>(&cs);
+
+        let transcript = Tr::new(&transcript_config);
```

Comment:
Should we allow to seed the transcript? It's usually done in  most of the proving systems. But up to you. Not mandatory for now at all.

Reply:
I just realized that we don't need the actual Transcript but just to use the Poseidon as RO, this is what it is already implemented, just that I was initializing the transcript but never using it. Just removed the transcript from the IVC code.
But agree, for other places where we will use the transcript (eg. in ProtoGalaxy) I agree to allow for setting the seed.